spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java

```diff
@@ -107,6 +108,8 @@ public class Mqttv5PahoMessageDrivenChannelAdapter
 
 	private volatile boolean readyToSubscribeOnStart;
 
+	private final AtomicInteger subIdCounter = new AtomicInteger(0);
```

Comment:
I'm still not sure in this incrementing approach.
Let's imaging we have several `Mqttv5PahoMessageDrivenChannelAdapter` instances!
Each of them with its own client id, but still same set of topics to subscribe.
However all of them are started with the same initial count.
So, different instances on the same topics might get same identifiers.

Now another scenario: we have several instances in the cluster with the same `clientId`, but according to their different JVMs they might get different identifiers.

So, how it is really supposed to work correctly if there is no correlation between subscriber instance and that identifier?

I need more proves that a plain counter which produces same number for different instances is really what is enough for those shared subscriptions to work.

Thanks

Reply:
Subscription identifier is meaningful only to the client that does the SUBSCRIBE operation. 

From the spec 3.8.2.1.2 Subscription Identifier: _[..]The Subscription Identifier is associated with any subscription created or modified as the result of this SUBSCRIBE packet. If there is a Subscription Identifier, it is stored with the subscription. If this property is not specified, then the absence of a Subscription Identifier is stored with the subscription._

From the spec 3.8.4 SUBSCRIBE Actions: _The Subscription Identifiers are part of the Session State in the Server and are returned to the Client receiving a matching PUBLISH packet. They are removed from the Server’s Session State when the Server receives an UNSUBSCRIBE packet, when the Server receives a SUBSCRIBE packet from the Client for the same Topic Filter but with a different Subscription Identifier or with no Subscription Identifier, or when the Server sends Session Present 0 in a CONNACK packet._

Reply:
I'm sorry, I'm still failing to understand that description.

Would you mind to rephrase it some how to confirm that plain counter is enough for us for any use-cases I described before?

Thanks

Reply:
The subscription identifiers are part of the server side session state for the client. From [4.1 Session State](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Session_State):

> The Session State in the Server consists of:
> [..]
> The Clients subscriptions, including any Subscription Identifiers.

So the scope of Subscription identifiers is the session.

> Let's imaging we have several Mqttv5PahoMessageDrivenChannelAdapter instances!
> Each of them with its own client id, but still same set of topics to subscribe.
> However all of them are started with the same initial count.
> So, different instances on the same topics might get same identifiers.

If they add the topics in the same order each then they will generate the same Subscription identifiers. But that is fine because those Subscription identifiers are per session.

> Now another scenario: we have several instances in the cluster with the same clientId, but according to their different JVMs they might get different identifiers.

You cannot have the same client ID for multiple connections. If the server gets a connection with the same client ID, it has to terminate the existing connection and use the new connection. From [3.1.4 CONNECT Actions](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901073):

> [..]If the ClientID represents a Client already connected to the Server, the Server sends a DISCONNECT packet to the existing Client with Reason Code of 0x8E (Session taken over) as described in [section 4.13](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#S4_13_Errors) and MUST close the Network Connection of the existing Client.

[ActiveMQ Artemis](https://activemq.apache.org/components/artemis/documentation/latest/mqtt.html#link-stealing) allows to change this and reject the new connection in favour of the existing one.

> I need more proves that a plain counter which produces same number for different instances is really what is enough for those shared subscriptions to work.

It works because the Subscription identifier is per session and because you cannot have multiple connection with the same client ID.

Reply:
Sounds good!

Reviewing, merging and back-porting then.

Thank you!

Reply:
Since we do an Open Source here, it is better to be as readable as possible.
Therefore we prefer more clear names even for local variables.
My request here is to make this property as `subscriptionIdentifierCounter`

Reply:
Fixed.